lib/charms/ip_router_interface/v0/ip_router_interface.py

```diff
@@ -449,6 +451,46 @@ def request_network(self, networks: List[Network], custom_network_name: str = No
             str([r.name for r in ip_router_relations]),
         )
 
+    def get_routing_table(self) -> RoutingTable:
+        """Fetches combined routing tables made available by ip-router providers
+
+        Returns:
+            A list of objects of type `Network`. This list contains networks
+            from all ip-router providers that are integrated with the charm.
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        router_relations = self.model.relations.get(self.relationship_name)
+        validated_routing_table: RoutingTable = {}
+        for relation in router_relations:
+            if networks := relation.data[relation.app].get("networks"):
+                routing_table_from_databag: RoutingTable = json.loads(networks)
+                if not isinstance(json.loads(networks), dict):
+                    logger.error(
+                        "The router's routing table has been misconfigured. Can't build routing table."
+                    )
+                    return {}
+                for network_name, networks in routing_table_from_databag.items():
+                    validated_routing_table[network_name] = []
+                    for network in networks:
+                        try:
+                            _validate_network(network, validated_routing_table)
+                        except (ValueError, KeyError) as e:
+                            logger.warning(
+                                "Malformed network detected in the databag:\nNetwork: (%s)\nError: (%s)",
+                                network,
+                                e.args[0],
+                            )
+                        else:
+                            validated_routing_table[network_name].append(network)
+                logger.debug(
+                    f"Read networks from app: (%s) and relation: (%s)",
```

Comment:
```suggestion
                    "Read networks from app: (%s) and relation: (%s)",
```

Reply:
We want to avoid using fstrings in logging. Here we're not quite using them it seems though. For reference: https://docs.google.com/document/d/1H5W2oi7cuTaz86taTQiITv49jpJNKzzDZ2G25IQpE4w/edit#heading=h.r5obiqmhwoy3

Reply:
Yea i was aware of that, looks like the f slipped through some copy-paste cracks. Committed.